### PR TITLE
ensure cleanup after the gatsby plugin runs

### PR DESCRIPTION
**Pull request by EndangeredMassa**
There is an issue where you can interrupt a gatsby project build and the temporary files are not cleaned up. If you try to build again, it will get stuck in an infinite loop.

This PR adds an exit hook to make sure the cleanup happens.

### CODE REVIEW

File: packages/static-build/src/utils/gatsby.ts [hunk 1]
```diff
@@ -311,32 +321,39 @@ module.exports = gatsbyNode;
...
-        fs.readFile(backupFilePath, 'utf-8').catch(() => null),
-      ]);
-
-      if (
-        baseFileContent &&
-        baseFileContent.startsWith(GENERATED_FILE_COMMENT)
-      ) {
-        await fs.rm(baseFilePath);
+// must remain sync because it's called in an exit handler
+export function cleanupGatsbyFiles(dir: string) {
```
Comment (EndangeredMassa) [inline packages/static-build/src/utils/gatsby.ts:325]: This had to be made sync to be executable in a process exit hook.

File: packages/static-build/src/utils/gatsby.ts [hunk 2]
```diff
@@ -33,6 +36,13 @@ export async function injectPlugins(
     const version = semver.coerce(detectedVersion);
     if (version && semver.satisfies(version, '>=4.0.0')) {
       plugins.add('@vercel/gatsby-plugin-vercel-builder');
+
+      if (!GLOBAL_EXIT_HANDLER) {
+        GLOBAL_EXIT_HANDLER = () => {
+          cleanupGatsbyFiles(dir);
+        };
+        process.on('exit', GLOBAL_EXIT_HANDLER);
+      }
```
Comment (EndangeredMassa) [inline packages/static-build/src/utils/gatsby.ts:45]: I don't like adding exit hooks deep inside some logic, but we do try to make sure it's cleaned up later. If anyone has a suggestion on a better way to do this, please let me know.

File: packages/static-build/src/utils/gatsby.ts [hunk 3]
```diff
@@ -353,3 +370,19 @@ export async function createPluginSymlinks(dir: string) {
...
+  try {
+    return fs.readFileSync(filePath, 'utf-8');
+  } catch (error: unknown) {
+    if (!isErrnoException(error)) {
+      console.error(error);
+    } else {
+      if (error.code !== 'ENOENT') {
+        console.error(error);
+      }
+    }
```
Comment (TooTallNate) [inline packages/static-build/src/utils/gatsby.ts]:
```suggestion
    if (!isErrnoException(error) || error.code !== 'ENOENT') {
      console.error(error);
    }
```

**Review by TooTallNate — DISMISSED** [review]

**Review by Ethan-Arrowood — DISMISSED** [review]

File: packages/static-build/src/utils/gatsby.ts (hunk 2 above)
Comment (cb1kenobi) [reply packages/static-build/src/utils/gatsby.ts:45]: Maybe the files could be cleaned up before building?

File: packages/static-build/src/utils/gatsby.ts (hunk 3 above)
Comment (EndangeredMassa) [reply packages/static-build/src/utils/gatsby.ts]: I think this is clearer: [`f6d2c12` (#9975)](https://github.com/vercel/vercel/pull/9975/commits/f6d2c121fb0e5b4d0645cf74e24078c30e674b49)

**Review by Ethan-Arrowood — APPROVED** [review]

File: packages/static-build/src/utils/gatsby.ts (hunk 2 above)
Comment (EndangeredMassa) [reply packages/static-build/src/utils/gatsby.ts:45]: We really don't want to leave the user's project code in a temporary state. I think cleaning them up in process is best.

**Pull request merged**